source/common/upstream/load_balancer_impl.cc

```diff
@@ -267,6 +275,23 @@ bool ZoneAwareLoadBalancerBase::earlyExitNonLocalityRouting() {
   return false;
 }
 
+HostConstSharedPtr LoadBalancerBase::chooseHost(LoadBalancerContext* context) {
+  HostConstSharedPtr host;
+  const size_t max_attempts = context ? context->hostSelectionRetryCount() + 1 : 1;
+  for (size_t i = 0; i < max_attempts; ++i) {
+    host = chooseHostOnce(context);
+
+    // If host selection failed or the host is accepted by the filter, return.
+    // Otherwise, try again.
+    if (!host || !context || !context->shouldSelectAnotherHost(*host)) {
```

Comment:
I think the comment thread here was lost. I could see a scenario in which eventually a LB might not return nullptr on a subsequent attempt, in which case we might want to remove the `!host` check, but I agree with you this is fine for now and we can cross that bridge if we come to it.

Reply:
Sounds reasonable. Should I just add a TODO or file an issue?

Reply:
I would just add a comment of some type if you don't mind. I don't think it even needs to be a TODO but whatever you think is best.